src/Lykke.SettingsReader/SettingsProcessor/SettingsProcessor.cs

```diff
@@ -217,41 +231,68 @@ private static string CheckProperty<T>(PropertyInfo property, T model)
-                        if (optionalAttribute == null)
-                            throw new CheckFieldException(property.Name, val, "Empty setting value");
-                        continue;
-                    }
+                tasks = valuesToCheck.Select(val => DoCheck(model, property, checker, val)).ToList();
 
-                    var checkResult = checker.CheckField(model, property.Name, val);
-                    Console.WriteLine(checkResult.Description);
-                    if (!checkResult.Result && checkResult.ThrowExceptionOnFail)
-                        return checkResult.Description;
-                }
+                errorMessages = (await Task.WhenAll(tasks))
+                    .Where(item => !string.IsNullOrEmpty(item))
+                    .ToList();
+
+                return errorMessages.Count == 0 ? null : string.Join(Environment.NewLine, errorMessages);
             }
-            else if (property.CanWrite)
+
+            if (!property.CanWrite) 
+                return null;
+            
+            object[] values = GetValuesToCheck(property, model);
+
+            tasks = values.Select(ProcessChecks).ToList();
+
+            errorMessages = (await Task.WhenAll(tasks))
+                .Where(item => !string.IsNullOrEmpty(item))
+                .ToList();
+            
+            return errorMessages.Count == 0 ? null : string.Join(Environment.NewLine, errorMessages);
+        }
+
+        private static async Task<string> DoCheck<T>(T model, PropertyInfo property, ISettingsFieldChecker checker, string val)
+        {
+            if (string.IsNullOrWhiteSpace(val))
             {
-                object[] values = GetValuesToCheck(property, model);
-                var errorMessages = new List<string>();
-                Parallel.ForEach(values, val =>
-                {
-                    string errorMessage = ProcessChecks(val);
-                    if (errorMessage == null)
-                        return;
+                var optionalAttribute = property.GetCustomAttribute(typeof(OptionalAttribute));
+                
+                if (optionalAttribute == null)
+                    throw new CheckFieldException(property.Name, val, "Empty setting value");
+                
+                return null;
+            }
 
-                    lock (errorMessages)
-                    {
-                        errorMessages.Add(errorMessage);
-                    }
-                });
+            CheckFieldResult checkResult = await _retry.ExecuteAsync(() => Task.FromResult(checker.CheckField(model, property.Name, val)));
 
-                return errorMessages.Count == 0 ? null : string.Join(Environment.NewLine, errorMessages);
+            if (checkResult.Result) 
+                return null;
+            
+            await SendSlackNotification(checkResult.Description);
+            return checkResult.Description;
+        }
+
+        private static Task SendSlackNotification(string message)
```

Comment:
Let's use  Lykke.SlackNotifications.ISlackNotificationsSender instead of additional implementation of slack message sending

Reply:
no, DI container is not initialized here yet, so we need to create publisher etc by hand. That's why its implemented here this way

Reply:
You could use SlackNotificationsSender without DI container